src/Temporalio/Bridge/src/runtime.rs

```diff
@@ -0,0 +1,223 @@
+pub struct RuntimeOptions {
+    telemetry: *const TelemetryOptions,
+}
+
+#[repr(C)]
+pub struct TelemetryOptions {
+    tracing: *const TracingOptions,
+    logging: *const LoggingOptions,
+    metrics: *const MetricsOptions,
+}
+
+#[repr(C)]
+pub struct TracingOptions {
+    filter: ByteArrayRef,
+    opentelemetry: OpenTelemetryOptions,
+}
+
+#[repr(C)]
+pub struct LoggingOptions {
+    filter: ByteArrayRef,
+    forward: bool,
+}
+
+#[repr(C)]
+pub struct MetricsOptions {
+    opentelemetry: *const OpenTelemetryOptions,
+    prometheus: *const PrometheusOptions,
+}
+
+#[repr(C)]
+pub struct OpenTelemetryOptions {
+    url: ByteArrayRef,
+    /// Headers are <key1>\n<value1>\n<key2>\n<value2>. Header keys or values
+    /// cannot contain a newline within.
+    headers: ByteArrayRef,
+    metric_periodicity_millis: u32,
+}
+
+#[repr(C)]
+pub struct PrometheusOptions {
+    bind_address: ByteArrayRef,
+}
+
+#[derive(Clone)]
+pub struct Runtime {
+    pub(crate) core: Arc<CoreRuntime>,
+}
+
+/// If success or fail are not null, they must be manually freed when done.
+/// Runtime is always present, but it should never be used if fail is present,
+/// only freed after fail is freed.
+#[repr(C)]
+pub struct RuntimeOrFail {
+    runtime: *mut Runtime,
+    fail: *const ByteArray,
+}
+
+#[no_mangle]
+pub extern "C" fn runtime_new(options: *const RuntimeOptions) -> RuntimeOrFail {
+    match Runtime::new(unsafe { &*options }) {
+        Ok(runtime) => RuntimeOrFail {
+            runtime: Box::into_raw(Box::new(runtime)),
+            fail: std::ptr::null(),
+        },
+        Err(err) => {
+            // We have to make an empty runtime just for the failure to be
+            // freeable
+            let mut runtime = Runtime {
+                core: Arc::new(
+                    CoreRuntime::new(
+                        CoreTelemetryOptions::default(),
+                        tokio::runtime::Builder::new_current_thread(),
+                    )
+                    .unwrap(),
+                ),
+            };
+            let fail = runtime.alloc_utf8(&format!("Invalid options: {}", err));
+            RuntimeOrFail {
+                runtime: Box::into_raw(Box::new(runtime)),
+                fail: fail.into_raw(),
+            }
+        }
+    }
+}
+
+#[no_mangle]
+pub extern "C" fn runtime_free(runtime: *mut Runtime) {
+    unsafe {
+        let _ = Box::from_raw(runtime);
+    }
+}
+
+#[no_mangle]
+pub extern "C" fn byte_array_free(runtime: *mut Runtime, bytes: *const ByteArray) {
+    // Bail if freeing is disabled
+    unsafe {
+        if bytes.is_null() || (*bytes).disable_free {
+            return;
+        }
+    }
+    let bytes = bytes as *mut ByteArray;
+    // Return vec back to core before dropping bytes
+    let vec = unsafe { Vec::from_raw_parts((*bytes).data as *mut u8, (*bytes).size, (*bytes).cap) };
+    // Set to null so the byte dropper doesn't try to free it
+    unsafe { (*bytes).data = std::ptr::null_mut() };
```

Comment:
IMO this would be better accomplished by reconstituting the `ByteArray` and just dropping it.

Reply:
Not following exactly. Are you suggesting I `Box::from_raw` higher? I am using/updating all of the unsafe pointers before I return ownership back to Rust at the bottom.

Reply:
Did a little more digging and there's a function exactly for this: 

https://doc.rust-lang.org/std/ptr/fn.drop_in_place.html

Reply:
Not following the change exactly. I have to extract-and-remove the vec separately from the `ByteArray` so that I can give it back to the "runtime" (just in case we pool bytes later). Can you give a more concrete suggested change?

Reply:
Oh man, somehow I just totally missed the vec being passed back for re-use. Nevermind. :facepalm: